### ISSUE TITLE
Multiclass classification

### ISSUE THREAD

**Issue by AEugene** [KevinCoble/swift-libsvm#1]
hi @KevinCoble . Have you tried lib with multi-class classification? I got bad confusion matrix for Iris dataset.

**Comment by KevinCoble** [comment]
No, I have not.  The two-class case uses the same code (just shorter do-loops) as a multi-class, so I haven't dove further into it after getting a two-class example to work, instead going forward with the regression cases.  There are probably a few porting errors still, and I haven't got model persistence in yet either.

I'll try a simple case myself, but if you have an example that reads in a known data set and trains on it, I would be happy to use it to try and see what issues occur.

**Comment by KevinCoble** [comment]
I found a problem with the generation of the sub-problem data set when splitting multiple classes into two-class sets (also would have been a problem if a 2-case data set used non-standard labels)
I fixed the problem and added a 3-label test set to the XCTest file.  The result coefficients and support-vectors match the Windows pre-built LibSVM results with that test data set.

Hope this fixes your problem

**Comment by AEugene** [comment]
Tried it, but I'm not very satisfied with results. Here is the example https://github.com/AEugene/swift-libsvm/tree/feature/iris_dataset . @KevinCoble could you take a look? Maybe I'm doing something wrong.

**Comment by KevinCoble** [comment]
I got your example working and agree, the results don't look correct.  I am getting about a 30% match rate on the test data, which is about guess material.  I'll investigate further...

**Comment by KevinCoble** [comment]
I did much better when I fixed the error in your test-data scaling.  Look at the line that should be

```
                (iris.sepalWidth - minSepalWidth) / (maxSepalWidth - minSepalWidth),
```

You used the SepalLength parameters when scaling the test data (it is correctly using SepalWidth in the training data)

**Comment by KevinCoble** [comment]
I am still going to compare with the C++ version of LIBSVM, but I won't be able to till tomorrow (1/14)

**Comment by KevinCoble** [comment]
Removing scaling, training on all the data, and then testing on a random 25% (sort-of-cheating since the test data was also in the train data - but I need a fixed data set to compare with the C++ version), results in a near 97-100% success rate most runs.

**Comment by KevinCoble** [comment]
I ran the test data (all of it) through the pre-packaged LIBSVM on a Windows machine.  I compared the resulting model with that calculated by my Swift version.  While they differ, they are fairly close.  When debugging the initial port I found micro differences in the math libraries can lead to larger differences in the model (and I am using the Accelerate library with the vector processor, so there will be differences).  But the number of support vectors is similar for each machine, and most of the points listed as support vectors match.

I then ran all the data through the generated models to predict.  On the Windows system, two of the data points do not match.  On the Swift version 3 fail - with two of them being the same data points that the Windows system fails on.  This shows that the data is not linearly separable, at least with the kernel and cost values given.

You were training on half the data, then testing with the other half.  It is not surprising that a good number of the test failed, since there is obviously not a large margin between at least some of the data.

I think the code is correct, or at least is for this setup.  You may need to play with kernel parameters or costs to find the best separator for your data - although with only 150 data points, it is hard to use much of them for parameter searches and test data, and still leave enough to train on.
